File: src/artifact.ts
/**
 * Artifact profiles (spec §2.1): bytes/1 streams exact bytes; jcs/1 digests
 * B(J(value)) of a strictly parsed document (1 MiB raw cap); tree/1 digests
 * B(J(TreeManifest)) over a deterministic directory manifest.
 *
 * Tree capture rejects symlinks, multiply-hardlinked names, device files,
 * sockets, and filesystem changes during capture.  Files are hashed from
 * private staged copies; before/after stat checks detect ordinary races.
 */

import { createHash } from "node:crypto";
import {
  openSync, closeSync, readSync, fstatSync, lstatSync, readdirSync, statSync,
  mkdirSync, writeSync, fsyncSync,
} from "node:fs";
import { join, relative, sep } from "node:path";
import { SunlightError } from "./errors.js";
import { jcsBytes } from "./encoding/jcs.js";
import { parseStrictJson } from "./encoding/strict-json.js";
import { B } from "./crypto/domains.js";
import {
  parseTreeManifest, isValidTreePath,
  MAX_MANIFEST_BYTES, MAX_MANIFEST_FILES,
  type Artifact, type ArtifactProfile, type TreeFile, type TreeManifest,
} from "./schema.js";

const HASH_BUF = 4 * 1024 * 1024; // 4 MiB streaming cap (spec §10)
const JCS_RAW_CAP = 1024 * 1024;  // 1 MiB raw input cap for jcs/1 (spec §2.1)

/** Stream a single file's raw bytes through SHA-256 with a bounded buffer. */
export function hashFileBytes(path: string): { digest: string; bytes: number } {
  const fd = openSync(path, "r");
  try {
    const st = fstatSync(fd);
    if (!st.isFile()) throw new SunlightError("ARTIFACT_CHANGED");
    const h = createHash("sha256");
    const buf = Buffer.allocUnsafe(Math.min(HASH_BUF, Math.max(st.size, 4096)));
    let total = 0;
    for (;;) {
      const n = readSync(fd, buf, 0, buf.length, null);
      if (n === 0) break;
      h.update(buf.subarray(0, n));
      total += n;
    }
    return { digest: "sha256:" + h.digest("hex"), bytes: total };
  } finally {
    closeSync(fd);
  }
}

/** jcs/1 over an explicitly canonicalized JSON file. */
export function hashJcsFile(path: string): Artifact {
  const st = lstatSync(path);
  if (!st.isFile() || st.isSymbolicLink()) throw new SunlightError("ARTIFACT_CHANGED");
  if (st.size > JCS_RAW_CAP) throw new SunlightError("BODY_LIMIT");
  const raw = readFileBytes(path);
  let value: unknown;
  try {
    value = parseStrictJson(raw);
  } catch (e) {
    throw new SunlightError("SCHEMA_INVALID", null, e);
  }
  const canon = jcsBytes(value);
  return { profile: "jcs/1", digest: B(canon), bytes: canon.length };
}

function readFileBytes(path: string): Buffer {
  const fd = openSync(path, "r");
  try {
    const st = fstatSync(fd);
    const out = Buffer.allocUnsafe(st.size);
    let off = 0;
    while (off < st.size) {
      const n = readSync(fd, out, off, st.size - off, null);
      if (n === 0) break;
      off += n;
    }
    return out.subarray(0, off);
  } finally {
    closeSync(fd);
  }
}

interface FileSnap {
  dev: bigint;
  ino: bigint;
  size: bigint;
  mtimeNs: bigint;
}

function snap(path: string): FileSnap {
  const st = lstatSync(path, { bigint: true });
  return { dev: st.dev, ino: st.ino, size: st.size, mtimeNs: st.mtimeNs };
}

function sameSnap(a: FileSnap, b: FileSnap): boolean {
  return a.dev === b.dev && a.ino === b.ino && a.size === b.size && a.mtimeNs === b.mtimeNs;
}

/** Enumerate a directory into validated relative paths (source order). */
export function enumerateTree(root: string): string[] {
  const out: string[] = [];
  const walk = (dir: string, rel: string): void => {
    let names: string[];
    try {
      names = readdirSync(dir);
    } catch (e) {
      throw new SunlightError("IO_ERROR", null, e);
    }
    for (const name of names) {
      const full = join(dir, name);
      const r = rel === "" ? name : rel + "/" + name;
      const st = lstatSync(full);
      if (st.isSymbolicLink()) {
        throw new SunlightError("SCHEMA_INVALID", null, new Error(`symlink rejected: ${r}`));
      }
      if (st.isDirectory()) {
        walk(full, r);
        continue;
      }
      if (!st.isFile()) {
        throw new SunlightError("SCHEMA_INVALID", null, new Error(`non-regular file rejected: ${r}`));
      }
      if (st.nlink > 1) {
        throw new SunlightError("SCHEMA_INVALID", null, new Error(`multiply-linked file rejected: ${r}`));
      }
      if (!isValidTreePath(r)) {
        throw new SunlightError("SCHEMA_INVALID", null, new Error(`invalid tree path: ${r}`));
      }
      out.push(r);
    }
  };
  const rootStat = lstatSync(root);
  if (!rootStat.isDirectory() || rootStat.isSymbolicLink()) {
    throw new SunlightError("USAGE", null, new Error("tree root is not a directory"));
  }
  walk(root, "");
  out.sort();
  if (out.length > MAX_MANIFEST_FILES) {
    throw new SunlightError("BODY_LIMIT", null, new Error("tree over 10000 files"));
  }
  const fold = new Set<string>();
  for (const p of out) {
    const f = p.toLowerCase();
    if (fold.has(f)) throw new SunlightError("SCHEMA_INVALID", null, new Error(`case collision: ${p}`));
    fold.add(f);
  }
  return out;
}

/**
 * Stage a directory into `stagingDir` (created 0700 by caller), hash every
 * staged file, and produce the sorted manifest.  Copies into private staging
 * first so the hashed bytes are exactly the bytes downstream consumers see.
 */
export function stageTree(root: string, stagingDir: string): { manifest: TreeManifest; stagedFiles: Map<string, string> } {
  const rels = enumerateTree(root);
  const staged = new Map<string, string>();
  const before = new Map<string, FileSnap>();
  for (const rel of rels) {
    const src = join(root, ...rel.split("/"));
    const dst = join(stagingDir, ...rel.split("/"));
    mkdirSync(join(stagingDir, ...rel.split("/").slice(0, -1)), { recursive: true, mode: 0o700 });
    before.set(rel, snap(src));
    const data = readFileBytes(src);
    const fd = openSync(dst, "wx", 0o600);
    try {
      writeSync(fd, data);
      fsyncSync(fd);
    } finally {
      closeSync(fd);
    }
    staged.set(rel, dst);
    // Post-copy race check on the source handle identity.
    const after = snap(src);
    if (!sameSnap(before.get(rel)!, after)) {
      throw new SunlightError("ARTIFACT_CHANGED", null, new Error(`file changed during capture: ${rel}`));
    }
  }
  const files: TreeFile[] = rels.map((rel) => {
    const { digest, bytes } = hashFileBytes(staged.get(rel)!);
    return { path: rel, digest, bytes };
  });
  const manifest = parseTreeManifest({ v: "sunlight.tree/1", files });
  if (jcsBytes(manifest).length > MAX_MANIFEST_BYTES) {
    throw new SunlightError("BODY_LIMIT");
  }
  return { manifest, stagedFiles: staged };
}

/** tree/1 artifact for a manifest value already built from staged files. */
export function manifestArtifact(manifest: TreeManifest): Artifact {
  const canon = jcsBytes(manifest);
  return { profile: "tree/1", digest: B(canon), bytes: canon.length };
}

/**
 * hashArtifact: bytes/1 streams the file; jcs/1 parses and canonicalizes;
 * tree/1 stages + manifests a directory.
 */
export async function hashArtifact(input: { profile: ArtifactProfile; path: string }): Promise<Artifact> {
  const { profile, path } = input;
  if (profile === "bytes/1") {
    const st = lstatSync(path);
    if (!st.isFile() || st.isSymbolicLink()) {
      throw new SunlightError("USAGE", null, new Error("artifact path is not a regular file"));
    }
    const before = snap(path);
    const r = hashFileBytes(path);
    const after = snap(path);
    if (!sameSnap(before, after)) throw new SunlightError("ARTIFACT_CHANGED");
    return { profile: "bytes/1", digest: r.digest, bytes: r.bytes };
  }
  if (profile === "jcs/1") {
    return hashJcsFile(path);
  }
  // tree/1 — stage into a private temp dir alongside the source, hash staged.
  const { mkdtempSync, rmSync, chmodSync } = await import("node:fs");
  const { tmpdir } = await import("node:os");
  const staging = mkdtempSync(join(tmpdir(), "sunlight-tree-"));
  chmodSync(staging, 0o700);
  try {
    const { manifest } = stageTree(path, staging);
    return manifestArtifact(manifest);
  } finally {
    rmSync(staging, { recursive: true, force: true });
  }
}

File: python/sunlight/artifact.py
"""Artifact profiles (spec §2.1): bytes/1, jcs/1, tree/1.

Tree enumeration rejects symlinks, multiply-hardlinked names, non-regular
files, invalid/traversal paths, and case-insensitive collisions; manifests
sort by path before hashing.  All hashing is bounded and streaming.
"""

from __future__ import annotations

import hashlib
import os
import tempfile

from .domains import B
from .errors import SunlightError
from .jcs import jcs
from .schema import (
    MAX_MANIFEST_BYTES, MAX_MANIFEST_FILES, is_valid_tree_path, parse_tree_manifest,
)
from .strictjson import parse_strict_json

_HASH_BUF = 4 * 1024 * 1024
_JCS_RAW_CAP = 1024 * 1024


def hash_file_bytes(path: str) -> dict:
    st = os.lstat(path)
    if not os.path.isfile(path) or os.path.islink(path):
        raise SunlightError("ARTIFACT_CHANGED", cause=ValueError("not a regular file"))
    h = hashlib.sha256()
    total = 0
    with open(path, "rb") as f:
        while True:
            chunk = f.read(_HASH_BUF)
            if not chunk:
                break
            h.update(chunk)
            total += len(chunk)
    return {"digest": "sha256:" + h.hexdigest(), "bytes": total}


def _snap(path: str):
    st = os.lstat(path)
    return (st.st_dev, st.st_ino, st.st_size, st.st_mtime_ns)


def hash_jcs_file(path: str) -> dict:
    st = os.lstat(path)
    if not os.path.isfile(path) or os.path.islink(path):
        raise SunlightError("ARTIFACT_CHANGED")
    if st.st_size > _JCS_RAW_CAP:
        raise SunlightError("BODY_LIMIT")
    with open(path, "rb") as f:
        raw = f.read()
    canon = jcs(parse_strict_json(raw))
    return {"profile": "jcs/1", "digest": B(canon), "bytes": len(canon)}


def enumerate_tree(root: str) -> list[str]:
    st = os.lstat(root)
    if not os.path.isdir(root) or os.path.islink(root):
        raise SunlightError("USAGE", cause=ValueError("tree root is not a directory"))
    out: list[str] = []
    stack = [(root, "")]
    while stack:
        d, rel = stack.pop()
        for name in os.listdir(d):
            full = os.path.join(d, name)
            r = name if rel == "" else rel + "/" + name
            st = os.lstat(full)
            if os.path.islink(full):
                raise SunlightError("SCHEMA_INVALID", cause=ValueError(f"symlink rejected: {r}"))
            if os.path.isdir(full):
                stack.append((full, r))
                continue
            if not os.path.isfile(full):
                raise SunlightError("SCHEMA_INVALID", cause=ValueError(f"non-regular file: {r}"))
            if st.st_nlink > 1:
                raise SunlightError("SCHEMA_INVALID", cause=ValueError(f"multiply-linked file: {r}"))
            if not is_valid_tree_path(r):
                raise SunlightError("SCHEMA_INVALID", cause=ValueError(f"invalid tree path: {r}"))
            out.append(r)
    out.sort()
    if len(out) > MAX_MANIFEST_FILES:
        raise SunlightError("BODY_LIMIT")
    fold = set()
    for p in out:
        f = p.lower()
        if f in fold:
            raise SunlightError("SCHEMA_INVALID", cause=ValueError(f"case collision: {p}"))
        fold.add(f)
    return out


def stage_tree(root: str, staging_dir: str):
    """Copy the tree into private staging, hashing the staged copies."""
    rels = enumerate_tree(root)
    files = []
    for rel in rels:
        src = os.path.join(root, *rel.split("/"))
        dst = os.path.join(staging_dir, *rel.split("/"))
        os.makedirs(os.path.dirname(dst), exist_ok=True)
        before = _snap(src)
        fd = os.open(dst, os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o600)
        try:
            with open(src, "rb") as f:
                while True:
                    chunk = f.read(_HASH_BUF)
                    if not chunk:
                        break
                    os.write(fd, chunk)
            os.fsync(fd)
        finally:
            os.close(fd)
        if _snap(src) != before:
            raise SunlightError("ARTIFACT_CHANGED", cause=ValueError(f"changed during capture: {rel}"))
        files.append({"path": rel, **hash_file_bytes(dst)})
    manifest = parse_tree_manifest({"v": "sunlight.tree/1", "files": files})
    if len(jcs(manifest)) > MAX_MANIFEST_BYTES:
        raise SunlightError("BODY_LIMIT")
    return manifest


def manifest_artifact(manifest) -> dict:
    canon = jcs(manifest)
    return {"profile": "tree/1", "digest": B(canon), "bytes": len(canon)}


def hash_artifact(path: str, profile: str) -> dict:
    if profile == "bytes/1":
        st = os.lstat(path)
        if not os.path.isfile(path) or os.path.islink(path):
            raise SunlightError("USAGE", cause=ValueError("not a regular file"))
        before = _snap(path)
        r = hash_file_bytes(path)
        if _snap(path) != before:
            raise SunlightError("ARTIFACT_CHANGED")
        return {"profile": "bytes/1", "digest": r["digest"], "bytes": r["bytes"]}
    if profile == "jcs/1":
        return hash_jcs_file(path)
    if profile == "tree/1":
        staging = tempfile.mkdtemp(prefix="sunlight-tree-")
        try:
            os.chmod(staging, 0o700)
            return manifest_artifact(stage_tree(path, staging))
        finally:
            import shutil

            shutil.rmtree(staging, ignore_errors=True)
    raise SunlightError("USAGE", cause=ValueError("unknown profile"))
